cs/cs/Serializer/VowpalWabbitJsonBuilder.cs

```diff
@@ -0,0 +1,389 @@
+                    break;
+                default:
+                    // forward to handler
+                    if (specialPropertyAction == null || !specialPropertyAction(propertyName))
+                        reader.Skip(); // if not handled, skip it
+                    break;
+            }
+        }
+
+        private void ParseLabel()
+        {
+            // peak the first property name
+            if (!reader.Read())
+                throw new VowpalWabbitJsonException(reader.Path, "Unexpected end");
+
+            switch (reader.TokenType)
+            {
+                case JsonToken.StartObject:
+                    {
+                        // parse complex object
+                        if (!reader.Read() || reader.TokenType != JsonToken.PropertyName)
+                            throw new VowpalWabbitJsonException(reader.Path, "Expected at least a single property to determine the label object");
+
+                        var propertyName = (string)reader.Value;
+
+                        var prefixReader = new PrefixedJsonReader(this.reader,
+                            Tuple.Create(JsonToken.StartObject, (object)null),
+                            Tuple.Create(JsonToken.PropertyName, (object)propertyName));
+
+                        Type labelType;
+                        if (!labelPropertyMapping.TryGetValue(propertyName, out labelType))
+                            throw new VowpalWabbitJsonException(reader.Path, "The first property ('" + propertyName + "') must match to a property of a VowpalWabbit label type.");
+
+                        var label = (ILabel)jsonSerializer.Deserialize(prefixReader, labelType);
+
+                        this.defaultMarshaller.MarshalLabel(this.Context, label);
+                    }
+                    break;
+                case JsonToken.Integer:
+                case JsonToken.Float:
+                case JsonToken.String:
+                    {
+                        // pass label directly to VW
+                        var labelString = reader.Value.ToString();
+
+                        this.Context.ExampleBuilder.ParseLabel(labelString);
+                        // prefix with label
+                        this.Context.AppendStringExample(false, "{0}", labelString);
+                    }
+                    break;
+                default:
+                    throw new VowpalWabbitJsonException(reader.Path, "Expected label object");
+            }
+        }
+
+        /// <summary>
+        /// Expects: "1,2.2,3]" (excluding the leading [)
+        /// </summary>
+        private void ParseFeatureArray(VowpalWabbitMarshalContext context, Namespace ns)
+        {
+            ulong index = 0;
+
+            while (reader.Read())
+            {
+                switch (reader.TokenType)
+                {
+                    case JsonToken.Integer:
+                        MarshalFloatFeature(context, ns, index, (float)(long)reader.Value);
```

Comment:
reader.Value is of type object, the actual object is a long. Cast directly from object to float fails. The first cast gets us from object to long, the second actually does a conversion.